display: Give users the option to translate emoticons into emoji.

skunkmb opened the pull request:
This adds an option, `Translate emoticons`, in the display settings, which lets user have emoticons like `:)`, `<3`, `:(` and more translated into emoji like :smiley:, :heart:, :slightly_frowning_face:, etc. in their messages.

 1. A proprety `translate_emoticons` is added to the user model.
 2. New functions are added to `static/js/emoji.js` and `zerver/lib/emoji.py` which can convert emoticons into the emoji.
 3. The translator is added as a Markdown preprocessor in `markdown.js` for immediately after a message is sent and in `compose.js` for showing the Markdown preview.
 3. The translator is also used on the backend in `check_message` for translating emoticons whenever a user with the setting on sends a message.
 4. Add a setting in the `Display settings` for turning this feature on or off.

rht commented:
How much does this resemble #7931? The PR is too old that it is hard to compare which implementation is simpler.

skunkmb commented:
@rht I'm not sure how similar they are, but my PR is 404 LOC less than that one (222 LOC on mine and 626 on that PR, only looking at the last commit on the other PR because the other ones seem like they were added by accident). 

Another main difference is that my PR gives the setting to the **reader** of any message with an emoticon, whereas that PR seems to give the setting to the **writer** as to whether or not they want to translate emoticons in their messages.

aero31aero commented on static/js/emoji.js:
```diff
@@ -18,6 +18,36 @@ var zulip_emoji = {
+    ':( |-)?\\(': '🙁',
+    // Matches ":,(", ":, (", ":,-(", ":'(", ":' (", and ":'-(".
+    ':(,|\')( |-)?\\(': '😢',
+    // Matches ":O", ": O", ":-O", ":0", ": 0", and ":-0".
```
I would like to raise a point against this conversion. It is likely to have code snippets that have `: 0` as a part of them. On the whole, I do not see why we should translated emojis when the text has spaces in between. `: )` might be someone deliberately trying to avoid such a conversion.

timabbott replied:
Yeah, I agree; I think we should try to make this list minimal, to avoid accidental use of emoticons.  Avoiding accidental use of syntax is one of the philosophical principles in Zulip's message formatting, and it seems like we should stick with that here.

skunkmb replied:
@timabbott OK, I removed the space option in the RegExes. Please ket me know if there is more I should fix or if this could be merged. Thank you!

skunkmb replied:
(also @aero31aero)

rishig commented:
I also think it makes sense for this to be controlled on the writers side, and then to add this to the regular markdown pathways.

rishig commented:
@skunkmb, would you be up for taking over #7931, and finishing that? I think the feature that most people are looking for is the one on the writer's side.

skunkmb commented:
@rishig OK, I reworked it so that it now is on the writer's side. There is frontend JS for rendering the previews and the Markdown echo, but other than that the translation happens when the writer hits send, on the backend with Python.

rishig commented:
Awesome, thanks! A few quick observations:
* Likely the first two commits should be squashed into one, so that each commit is individually correct. Maybe with commit message: `emoji: Add support for translating emoticons.`
* I think we'll want to do the backend translation in `zerver/lib/bugdown`, with the rest of the markdown code. I don't know where the frontend markdown code is, but likely same story there. https://zulip.readthedocs.io/en/1.7.1/markdown.html might be relevant.
* One question (maybe for chat.zulip.org) is how generous to be in the conversion; e.g. how many emoticons to allow, and how many variants to allow in each one. There are many reasonable emoticons, e.g. https://get.slack.help/hc/en-us/articles/202931348-Emoji-and-emoticons#use-emoticons, https://aprescott.com/posts/hangouts-emoji, and some of the ones at https://en.wikipedia.org/wiki/List_of_emoticons.

Some arguments against allowing too many are: 
* markdown performance (not sure how big a deal that is)
* if the main use case is that people have some particular emoji they use a lot, like :smile:, and they just want a fast way to type it, then having a lot of variants of that doesn't add much value
* more emoticons -> more chance of false positives
* a smaller list makes reading the documentation for it less overwhelming

skunkmb commented:
@rishig Thanks! In regards to your comments:

 - OK, that makes sense. I just rebased and squashed them into one. Thanks!

 - I could be wrong, but I feel as though it makes more sense and is possibly more efficient to just 
    1. translate messages with emoticons (*only if their sender has the setting on, of course*) into their Markdown syntax (e.g. `:smiley:`) on send, and then
    2. treat them like normal Markdown later

   instead of translating them every time they are rendered with the other Markdown code. 

   That way we know that 

    - every message, translated-or-not, should be treated the same by Markdown, and it also means that 
    - messages won't become de-translated after people turn off the setting if they already sent some messages.

   That's the way I am currently doing it, and it seems to be working pretty well. I could be wrong, though, so please let me know if the other way makes more sense.

 - OK, I opened a topic in the #design stream: https://chat.zulip.org/#narrow/stream/design/subject/Emoticons.20for.20Translations/near/458295.

Please let me know if there is more I should change, and also your thoughts on bullet point # 2.

rishig commented:
I think maybe one misunderstanding here is that the markdown is only ever run once (well, each markdown processor is run once) on any given message. Message.content stores what the user typed, and Message.rendered_content stores the post-markdown html. 

The implementation here doesn't have to change, it just has to be moved into the markdown processor. I.e. I expect we could still implement this as "first, `:)` is converted to `:smile:`, then the emoji mappings are applied" in both markdown processors. 

One issue with doing message translation outside of the markdown processor is that it may break future code that assumes all translation is done in markdown. For instance, the /remind feature may create the Message object by calling the markdown directly, and then send the message later.

skunkmb commented:
@rishig Ah, that makes sense. Thanks! OK, I moved it into a Markdown Preprocessor now called [`EmoticonTranslationPreprocessor `](https://github.com/zulip/zulip/pull/8081/files#diff-237f2d7c30a5fdd38c7f1e2d8f6734eeR1173). There is also an emoticon-translating JS preprocessor, `preprocess_translate_emoticons` (although that was part of the PR before this new change). Please let me know if there is more I should change or fix. Thanks!

rishig commented:
cool, thanks @skunkmb! Next steps here are probably to figure out the list of emoticons to accept. By default I would start with most of the slack ones, and see if there are any we can add for our list of 6 common emoji 
![image](https://user-images.githubusercontent.com/890911/35318090-b666930e-008e-11e8-93bb-78b738e0ccbd.png)
I would not include the ones that are `:[a-z].*`, and for `:D`, I would make sure our typeahead does a reasonable thing.

Some other notes:
* Make sure there is a test ensuring that things in `code fences` don't get emojified, and any other corner cases you can think of.
* Barring thoughts from czo, I would just do lists of strings, rather than regular expressions, both for readability and lower chance of regressions/making a mistake in the future.

timabbott commented on zerver/views/user_settings.py:
```diff
@@ -132,6 +132,7 @@ def update_display_settings_backend(
         twenty_four_hour_time: Optional[bool]=REQ(validator=check_bool, default=None),
```
`zerver/views/__pycache__/pointer.cpython-34.pyc.140398671546048` should not be checked in.

timabbott commented on zerver/lib/bugdown/__init__.py:
```diff
@@ -1596,6 +1610,10 @@ def extendMarkdown(self, md: markdown.Markdown, md_globals: Dict[str, Any]) -> N
 
+        md.preprocessors.add('translate_emoticons',
+                             EmoticonTranslationPreprocessor(md),
+                             '_begin')
```
I'm not usre `_begin` is the right place in the ordering for this to go, but it might be.  The way to tell easily is by having tests for potential ordering issues (e.g. putting the syntax inside various types of code blocks; emoticons shouldn't be translated in that context).

timabbott commented on zerver/lib/bugdown/__init__.py:
```diff
@@ -1169,6 +1170,19 @@ def __init__(self, parser: Any) -> None:
+        new_lines = []  # type: List[Text]
+
+        for line in lines:
+            new_lines.append(translate_emoticons(line))
```
It seems like this will only translate syntax on a line by itself; is that the desired behavior?

timabbott commented:
Responding to a few things @rishig said:

> Barring thoughts from czo, I would just do lists of strings, rather than regular expressions, both for readability and lower chance of regressions/making a mistake in the future.

What I'd do is a list of strings, and then use something like `re.escape` to convert those into a regular expression for inclusion in bugdown (we'll want a regex to handle issues like requiring surrounding whitespace-or-start/end-of-message/line for the syntax to trigger).  

> Next steps here are probably to figure out the list of emoticons to accept.

I'd actually prefer to just try to get merged a correct implementation with at most the ~6 we have, and then do a follow-up commit series for adding the rest.  Just as a way of limiting growth in complexity of this PR.  Especially given that we're still talking about how to encode them.

rishig commented:
> correct implementation with at most the ~6 we have

sounds good! Let's do `:), (:, :(, ):, <3, :|, :/` or a subset of those since I'm confident we'll want them eventually.

skunkmb commented:
@rishig OK, I

 - Added a `help` document for the feature,
 - Converted the list of RegExes to a list of strings that get escaped.
 - Made tests for the feature.

However, I can't figure out how to avoid translating emoji in `code blocks` and was wondering if you could give me some guidance as to where to look. I was able to successfully get code *fences* to be avoided, based on code-fence-avoider in `BugdownUListPreprocessor`, but I'm not sure how to avoid inline code blocks. Any advice would be appreciated. Thanks in advance!

timabbott commented:
@skunkmb for code blocks, the way to do that is usually to put the third argument to e.g. `md.foo.bar.add` as something like  `'>backtick'` (i.e. it's processed after code blocks have turned things inside them into atomic objects), not `_begin`.

skunkmb commented:
@rishig @timabbott OK, thanks! The emoticon translation is now done with an [inline processor](https://github.com/zulip/zulip/pull/8081/files#diff-237f2d7c30a5fdd38c7f1e2d8f6734eeR1002), so translation doesn't happen inside code blocks anymore.

timabbott commented on zerver/lib/emoji.py:
```diff
@@ -14,6 +14,31 @@
+    ':/': ':confused:',
+}
+
+possible_emoticons = EMOTICON_CONVERSIONS.keys()
```
Do we need this variable?

skunkmb replied:
@timabbott Yes, because it is used to create `emoticon_regex` which is in turn [used in the Markdown pattern](https://github.com/zulip/zulip/pull/8081/files/7b51a3a130f668e97816e6a008f9f453d76b4128#diff-237f2d7c30a5fdd38c7f1e2d8f6734eeR1594).

timabbott commented on zerver/tests/test_home.py:
```diff
@@ -155,6 +155,8 @@ def test_home(self) -> None:
             "subscriptions",
             "test_suite",
             "timezone",
+            "total_uploads_size",
```
This is a rebase bug, I think; the other feature here was removed.

skunkmb replied:
@timabbott OK, I removed this line. Thanks!

timabbott commented on templates/zerver/help/enable-emoticon-translations.md:
```diff
@@ -0,0 +1,24 @@
+# Enable emoticon translation
+
+If you use emoticons like `:)` or `:/`, you can have them translated into
+emoji equivalents like 😃 or 😕 automatically by Zulip.
```
Look at how our emoji are done in these docs; there might be a way to have this be an image.

skunkmb replied:
@timabbott OK, thanks! I changed it so that it uses an image now, like the way `add-emoji` does it. I also had to convert the Markdown table to HTML because I was getting linter errors.

timabbott commented on static/js/settings_display.js:
```diff
@@ -184,6 +184,30 @@ exports.set_up = function () {
+            context.new_mode = 'be';
+        } else {
+            context.new_mode = 'not be';
+        }
```
I think we need these strings tagged for translation?  This whole system is annoying and we're hoping to change, but I think that's correct.

skunkmb replied:
@timabbott OK, I rebased the commit and they are tagged for translation now.

timabbott commented on static/js/server_events_dispatch.js:
```diff
@@ -287,11 +287,13 @@ exports.dispatch_normal_event = function dispatch_normal_event(event) {
         }
-        if (event.setting_name === 'twenty_four_hour_time') {
+        if (event.setting_name === 'twenty_four_hour_time'
+            || event.setting_name === 'translate_emoticons') {
```
This doesn't require a rerender, since it only impacts future messages you send, not old ones.  I think the default block of updating `page_params` is all we need.

skunkmb replied:
@timabbott OK, I undid this change. I think it was a holdover from a different way I was doing it before. Thanks!

timabbott commented:
@skunkmb this looks pretty good; I posted a batch of comments (and you should also fix the test failures).

timabbott commented on zerver/fixtures/markdown_test_cases.json:
```diff
@@ -336,9 +336,21 @@
+      "name": "translate_emoticons_in_code",
+      "input":  "`:)`",
+      "expected_output": "<p><code>:)</code></p>",
+      "text_content": "`:)`"
```
The tests should go in whatever previous commit introduces what they're testing, rather than their own commit.

skunkmb replied:
@timabbott OK, I squashed the tests with the previous commit.

aero31aero commented:
Hey! Just a reminder in case you missed it. Your tests are still failing. Try fixing them? :)

HarshitOnGitHub commented on zerver/lib/emoji.py:
```diff
@@ -14,6 +14,30 @@
 
+# Emoticons and which emoji they should become. Duplicate emoji are allowed.
+# Changes here should be mimicked in `static/js/emoji.js`.
+EMOTICON_CONVERSIONS = {
```
@skunkmb Was just going through this PR and I think it would make more sense to export this mapping using `build_emoji` rather than hard coding it here and in `emoji.js`. That way they will not go out of sync. What do you think? Thanks for working on this! :)

timabbott replied:
Also, that's probably a good idea (to have the data structure be sent from server to client via that data structure).  @HarshitOnGitHub can you open an issue for doing that follow-up cleanup?  Not a high priority until we're changing the list, but seems useful.

timabbott replied:
@HarshitOnGitHub just checking in on this; might be a good quick thing for you to fix post release?

HarshitOnGitHub replied:
Sure, I am adding this to my ToDo! :)

timabbott commented:
@skunkmb this looks pretty good!  I decided to take a careful look through this.  I made a few changes around the whitespace behavior here:
```
diff --git a/zerver/fixtures/markdown_test_cases.json b/zerver/fixtures/markdown_test_cases.json
index a21215a..30a1cc8 100644
--- a/zerver/fixtures/markdown_test_cases.json
+++ b/zerver/fixtures/markdown_test_cases.json
@@ -347,6 +347,12 @@
       "text_content": ":) foo :( bar <3 with space : ) real emoji :smiley:"
     },
     {
+      "name": "translate_emoticons_whitepsace",
+      "input":  "a:) ;)b",
+      "expected_output": "<p>a:) ;)b</p>",
+      "text_content": "a:) ;)b"
+    },
+    {
       "name": "translate_emoticons_in_code",
       "input":  "`:)`",
       "expected_output": "<p><code>:)</code></p>",
diff --git a/zerver/lib/bugdown/__init__.py b/zerver/lib/bugdown/__init__.py
index bc6089a..4464e70 100644
--- a/zerver/lib/bugdown/__init__.py
+++ b/zerver/lib/bugdown/__init__.py
@@ -1008,9 +1008,9 @@ class EmoticonTranslation(markdown.inlinepatterns.Pattern):
             return None
 
         emoticon = match.group('emoticon')
-        emoji = markdown.util.etree.Element('span')
-        emoji.text = translate_emoticons(emoticon)
-        return emoji
+        translated = translate_emoticons(emoticon)
+        name = translated[1:-1]
+        return make_emoji(name_to_codepoint[name], translated)
 
 class UnicodeEmoji(markdown.inlinepatterns.Pattern):
     def handleMatch(self, match: Match[Text]) -> Optional[Element]:
@@ -1591,8 +1591,8 @@ class Bugdown(markdown.Extension):
             'tex',
             Tex(r'\B(?<!\$)\$\$(?P<body>[^\n_$](\\\$|[^$\n])*)\$\$(?!\$)\B'),
             '>backtick')
-        md.inlinePatterns.add('translate_emoticons', EmoticonTranslation(emoticon_regex), '_end')
         md.inlinePatterns.add('emoji', Emoji(EMOJI_REGEX), '_end')
+        md.inlinePatterns.add('translate_emoticons', EmoticonTranslation(emoticon_regex), '>emoji')
         md.inlinePatterns.add('unicodeemoji', UnicodeEmoji(unicode_emoji_regex), '_end')
         md.inlinePatterns.add('link', AtomicLinkPattern(markdown.inlinepatterns.LINK_RE, md), '>avatar')
 
diff --git a/zerver/lib/emoji.py b/zerver/lib/emoji.py
index bd4d68d..ca17d1a 100644
--- a/zerver/lib/emoji.py
+++ b/zerver/lib/emoji.py
@@ -27,7 +27,7 @@ EMOTICON_CONVERSIONS = {
 
 possible_emoticons = EMOTICON_CONVERSIONS.keys()
 possible_emoticon_regexes = map(re.escape, possible_emoticons)
-emoticon_regex = '(?P<emoticon>(' + ')|('.join(possible_emoticon_regexes) + '))'
+emoticon_regex = '(?<![^\s])(?P<emoticon>(' + ')|('.join(possible_emoticon_regexes) + '))(?![\S])'
 
 # Translates emoticons to their colon syntax, e.g. `:smiley:`.
 def translate_emoticons(text: Text) -> Text:
diff --git a/zerver/tests/test_bugdown.py b/zerver/tests/test_bugdown.py
index 5d3802c..fc5f440 100644
--- a/zerver/tests/test_bugdown.py
+++ b/zerver/tests/test_bugdown.py
@@ -624,7 +624,7 @@ class BugdownTest(ZulipTestCase):
         msg = Message(sender=user_profile, sending_client=get_client("test"))
 
         content = u':)'
-        expected = u':)'
+        expected = u'<p>:)</p>'
         converted = render_markdown(msg, content)
         self.assertEqual(converted, expected)

```

And I fixed all the failing tests.  This is now merged.  There are two immediate follow-up issues for further cleaning this up: #8585 and #8586.
